sdks/java/core/src/main/java/org/apache/beam/sdk/util/WatermarkHold.java

```diff
@@ -264,87 +260,140 @@ private Instant addElementHold(ReduceFn<?, ?, ?, W>.ProcessValueContext context)
+    Instant hold = addEndOfWindowHold(context, paneIsEmpty);
     if (hold == null) {
-      hold = addGarbageCollectionHold(context);
+      hold = addGarbageCollectionHold(context, paneIsEmpty);
     }
     return hold;
   }
 
   /**
-   * Add an end-of-window hold. Return the {@link Instant} at which hold was added,
-   * or {@literal null} if no hold was added.
-   *
-   * <p>The end-of-window hold guarantees that any empty {@code ON_TIME} pane can be given
-   * a timestamp which will not be considered beyond allowed lateness by any downstream computation.
+   * Attempt to add an 'end-of-window hold'. Return the {@link Instant} at which the hold was added
+   * (ie the end of window time), or {@literal null} if no hold was added. We only add the hold if
+   * we can be sure a timer will be set (by {@link ReduceFnRunner}) to clear it. In other words, the
+   * input watermark cannot be ahead of the end of window time.
+   * <p>An end-of-window hold is added in two situations:
+   * <ol>
+   * <li>An incoming element came in behind the output watermark (so we are too late for placing
+   * the usual element hold), but it may still be possible to include the element in an
+   * {@link Timing#ON_TIME} pane. We place the end of window hold to ensure that pane will
+   * not be considered late by any downstream computation.
+   * <li>We guarantee an {@link Timing#ON_TIME} pane will be emitted for all windows (except the
+   * {@link GlobalWindow}, see below) which saw at least one element, even if that
+   * {@link Timing#ON_TIME} pane is empty. Thus when elements in a pane are processed due to a
+   * fired trigger we must set both an end of window timer and an end of window hold. Again, the
+   * hold ensures the {@link Timing#ON_TIME} pane will not be considered late by any downstream
+   * computation.
+   * </ol>
+   * <p>If we are in the {@link GlobalWindow} then case 1 above still holds since during a
+   * pipeline drain the {@code ON_TIME} pane should be emitted if it is non-empty. However we
+   * declare case 2 is not supported.
+   * <p>We use {@code paneIsEmpty} to distinguish cases 1 and 2.
    */
   @Nullable
-  private Instant addEndOfWindowHold(ReduceFn<?, ?, ?, W>.Context context) {
-    // Only add an end-of-window hold if we can be sure a timer will be set to clear it
-    // by the end of window (ie the end of window is at or ahead of the input watermark).
+  private Instant addEndOfWindowHold(ReduceFn<?, ?, ?, W>.Context context, boolean paneIsEmpty) {
     Instant outputWM = timerInternals.currentOutputWatermarkTime();
     Instant inputWM = timerInternals.currentInputWatermarkTime();
 
     String which;
-    boolean tooLate;
+    boolean added;
     Instant eowHold = context.window().maxTimestamp();
     if (eowHold.isBefore(inputWM)) {
       which = "too late for end-of-window timer";
-      tooLate = true;
+      added = false;
+    } else if (paneIsEmpty && !eowHold.isBefore(GlobalWindow.INSTANCE.maxTimestamp())) {
```

Comment:
This should be explicitly managed as with final pane behavior. It should default to false.